File: data-structures/Python/stacks/stack_using_array_test.py
import unittest
from stacks import stacks_using_array as stack


class StackUsingArrayTest(unittest.TestCase):

    @classmethod
    def setUp(cls):
        stack.push(1)
        stack.push(2)
        stack.push(3)
        stack.push(4)

    def test_push(self):
        stack.push(5)
        self.assertEqual(5, stack.peek())
        stack.push(6)
        self.assertEqual(6, stack.peek())
        stack.push(7)
        self.assertEqual(7, stack.peek())
        stack.push(8)
        self.assertEqual(8, stack.peek())
        stack.push(9)
        self.assertEqual(9, stack.peek())

    def test_is_empty(self):
        self.assertEqual(False, stack.is_empty())
        stack.pop()
        self.assertEqual(False, stack.is_empty())
        stack.pop()
        self.assertEqual(False, stack.is_empty())
        stack.pop()
        self.assertEqual(False, stack.is_empty())
        stack.pop()
        self.assertEqual(True, stack.is_empty())
        stack.push(5)
        self.assertEqual(False, stack.is_empty())

    def test_pop(self):
        self.assertEqual(4, stack.pop())
        self.assertEqual(3, stack.pop())
        self.assertEqual(2, stack.pop())
        self.assertEqual(1, stack.pop())

    def test_peek(self):
        self.assertEqual(4, stack.peek())
        stack.push(5)
        self.assertEqual(5, stack.peek())
        stack.pop()
        stack.pop()
        self.assertEqual(3, stack.peek())



if __name__ == '__main__':
    unittest.main()

File: data-structures/Java/src/test/java/org/redquark/dsa/datastructures/stacks/StackUsingArrayTest.java
package org.redquark.dsa.datastructures.stacks;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;


class StackUsingArrayTest {

    private final StackUsingArray stackUsingArray = new StackUsingArray();

    @BeforeEach
    void setUp() {
        stackUsingArray.push(1);
        stackUsingArray.push(2);
        stackUsingArray.push(3);
        stackUsingArray.push(4);
    }

    @Test
    void testPush() {
        assertEquals(4, stackUsingArray.peek());
        stackUsingArray.push(5);
        assertEquals(5, stackUsingArray.peek());
        stackUsingArray.push(6);
        stackUsingArray.push(7);
        stackUsingArray.push(8);
        stackUsingArray.push(9);
        stackUsingArray.push(10);
        assertEquals(10, stackUsingArray.peek());
        assertThrows(StackOverflowError.class, () -> stackUsingArray.push(11));
    }

    @Test
    void testIsEmpty() {
        assertFalse(stackUsingArray.isEmpty());
        stackUsingArray.pop();
        stackUsingArray.pop();
        stackUsingArray.pop();
        stackUsingArray.pop();
        assertTrue(stackUsingArray.isEmpty());
        stackUsingArray.push(5);
        assertFalse(stackUsingArray.isEmpty());
        stackUsingArray.pop();
        assertTrue(stackUsingArray.isEmpty());
    }

    @Test
    void testPop() {
        assertEquals(4, stackUsingArray.pop());
        assertEquals(3, stackUsingArray.pop());
        assertEquals(2, stackUsingArray.pop());
        assertEquals(1, stackUsingArray.pop());
        assertThrows(RuntimeException.class, stackUsingArray::pop);
        stackUsingArray.push(5);
        assertEquals(5, stackUsingArray.pop());
        assertThrows(RuntimeException.class, stackUsingArray::pop);
    }

    @Test
    void testPeek() {
        assertEquals(4, stackUsingArray.peek());
        assertEquals(4, stackUsingArray.peek());
        stackUsingArray.pop();
        assertEquals(3, stackUsingArray.peek());
        stackUsingArray.pop();
        assertEquals(2, stackUsingArray.peek());
        stackUsingArray.pop();
        stackUsingArray.pop();
        assertThrows(RuntimeException.class, stackUsingArray::peek);
    }
}